ICE: E0121 `suggestion must not have overlapping parts`

<!--
[31mICE[0m: Rustc ./ice.rs '' 'assertion `left == right` failed: suggestion must not have overlapping parts   '  left: Some([SubstitutionPart { span: /tmp/im/ice.rs:3:17: 3:22 (#0), snippet: "T" }, SubstitutionPart { span: /tmp/im/ice.rs:3:21: 3:22 (#0), snippet: "T" }])' ' right: None'', 'assertion `left == right` failed: suggestion must not have overlapping parts   '  left: Some([SubstitutionPart { span: /tmp/im/ice.rs:3:17: 3:22 (#0), snippet: "T" }, SubstitutionPart { span: /tmp/im/ice.rs:3:21: 3:22 (#0), snippet: "T" }])' ' right: None''
File: /tmp/im/ice.rs
-->

auto-reduced (treereduce-rust):
````rust
struct _;

fn mainIterator<_ = _> {
    
}


````

original:
````rust
struct _;

fn mainIterator<_ = _> {
    let _: Box<Item = _> = Box::new();
}

````

Version information
````
rustc 1.79.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.79.0-dev
LLVM version: 18.1.3

````

Command:
`/home/matthias/.rustup/toolchains/local-debug-assertions/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error: expected identifier, found reserved identifier `_`
 --> /tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs:1:8
  |
1 | struct _;
  |        ^ expected identifier, found reserved identifier

error: expected identifier, found reserved identifier `_`
 --> /tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs:3:17
  |
3 | fn mainIterator<_ = _> {
  |                 ^ expected identifier, found reserved identifier

error: missing parameters for function definition
 --> /tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs:3:23
  |
3 | fn mainIterator<_ = _> {
  |                       ^ help: add a parameter list

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs:5:2
  |
5 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs`

error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions
 --> /tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs:3:17
  |
3 | fn mainIterator<_ = _> {
  |                 ^^^^^
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
  = note: `#[deny(invalid_type_param_default)]` on by default

thread 'rustc' panicked at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_errors/src/diagnostic.rs:910:9:
assertion `left == right` failed: suggestion must not have overlapping parts
  left: Some([SubstitutionPart { span: /tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs:3:17: 3:22 (#0), snippet: "T" }, SubstitutionPart { span: /tmp/icemaker_global_tempdir.gS8iEyHKDNMu/rustc_testrunner_tmpdir_reporting.aJYK66omiNGs/mvce.rs:3:21: 3:22 (#0), snippet: "T" }])
 right: None
stack backtrace:
   0:     0x7126a1f80a44 - std::backtrace_rs::backtrace::libunwind::trace::h265d1925ea598df2
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x7126a1f80a44 - std::backtrace_rs::backtrace::trace_unsynchronized::hb7395057f0c5468d
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7126a1f80a44 - std::sys_common::backtrace::_print_fmt::h3d47c991162d4e52
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7126a1f80a44 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hacda344c49f35c2d
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7126a1fc37e5 - core::fmt::rt::Argument::fmt::h73fccf37a3b29a55
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/fmt/rt.rs:142:9
   5:     0x7126a1fc37e5 - core::fmt::write::h5e0aa268252d3b32
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/fmt/mod.rs:1153:17
   6:     0x7126a1f4b6d9 - std::io::Write::write_fmt::hd8859b8fbc5fad9b
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/io/mod.rs:1843:15
   7:     0x7126a1f807fe - std::sys_common::backtrace::_print::hd162dc200d0a9165
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7126a1f807fe - std::sys_common::backtrace::print::h5331a7a997abba9a
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7126a1f6684a - std::panicking::default_hook::{{closure}}::hde322b8785a305f9
  10:     0x7126a1f66546 - std::panicking::default_hook::h8bb4e5b2167fa65c
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:291:9
  11:     0x7126a4b8e198 - rustc_driver_impl[3245b2fc008c51d3]::install_ice_hook::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_driver_impl/src/lib.rs:1351:17
  12:     0x7126a4b8e198 - <alloc[7a6344611ae1f7a8]::boxed::Box<rustc_driver_impl[3245b2fc008c51d3]::install_ice_hook::{closure#0}> as core[bf256ce745c7f201]::ops::function::Fn<(&dyn for<'a, 'b> core[bf256ce745c7f201]::ops::function::Fn<(&'a core[bf256ce745c7f201]::panic::panic_info::PanicInfo<'b>,), Output = ()> + core[bf256ce745c7f201]::marker::Send + core[bf256ce745c7f201]::marker::Sync, &core[bf256ce745c7f201]::panic::panic_info::PanicInfo)>>::call
                               at /home/matthias/vcs/github/rust_debug_assertions/library/alloc/src/boxed.rs:2032:9
  13:     0x7126a1f66dd1 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h8e32d665fbe06219
                               at /home/matthias/vcs/github/rust_debug_assertions/library/alloc/src/boxed.rs:2032:9
  14:     0x7126a1f66dd1 - std::panicking::rust_panic_with_hook::h6c21735fa2b9110c
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:792:13
  15:     0x7126a1f81cb6 - std::panicking::begin_panic_handler::{{closure}}::h34f9e231b5a18255
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:657:13
  16:     0x7126a1f80cc9 - std::sys_common::backtrace::__rust_end_short_backtrace::hc90aaf921d8b9378
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/sys_common/backtrace.rs:171:18
  17:     0x7126a1f669e7 - rust_begin_unwind
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/panicking.rs:645:5
  18:     0x7126a1fd9d56 - core::panicking::panic_fmt::hdfac44aca47052e0
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/panicking.rs:72:14
  19:     0x7126a1fda3ba - core::panicking::assert_failed_inner::h78cd3b39ce038bcf
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/panicking.rs:397:23
  20:     0x7126a5107f1f - core[bf256ce745c7f201]::panicking::assert_failed::<core[bf256ce745c7f201]::option::Option<&[rustc_errors[c9f9645da0d6d137]::SubstitutionPart; 2usize]>, core[bf256ce745c7f201]::option::Option<&[rustc_errors[c9f9645da0d6d137]::SubstitutionPart; 2usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/panicking.rs:357:5
  21:     0x7126a4ff53c7 - <rustc_errors[c9f9645da0d6d137]::diagnostic::Diag>::multipart_suggestion_with_style::<&str>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_errors/src/diagnostic.rs:910:9
  22:     0x7126a50179bf - <rustc_errors[c9f9645da0d6d137]::diagnostic::Diag>::multipart_suggestion::<&str>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_errors/src/diagnostic.rs:862:9
  23:     0x7126a50179bf - rustc_hir_analysis[578b254fcde5e3d0]::collect::placeholder_type_error_diag
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/collect.rs:237:17
  24:     0x7126a4f6edea - <dyn rustc_hir_analysis[578b254fcde5e3d0]::hir_ty_lowering::HirTyLowerer>::lower_fn_ty::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2403:28
  25:     0x7126a4f6edea - <dyn rustc_hir_analysis[578b254fcde5e3d0]::hir_ty_lowering::HirTyLowerer>::lower_fn_ty
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2327:5
  26:     0x7126a5029039 - rustc_hir_analysis[578b254fcde5e3d0]::collect::infer_return_ty_for_fn_sig
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/collect.rs:1418:17
  27:     0x7126a5029039 - rustc_hir_analysis[578b254fcde5e3d0]::collect::fn_sig
  28:     0x7126a6a38bb0 - rustc_query_impl[e2adab7f048175ff]::query_impl::fn_sig::dynamic_query::{closure#2}::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:624:47
  29:     0x7126a6a38bb0 - rustc_query_impl[e2adab7f048175ff]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2adab7f048175ff]::query_impl::fn_sig::dynamic_query::{closure#2}::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:548:18
  30:     0x7126a6cf18c5 - rustc_query_impl[e2adab7f048175ff]::query_impl::fn_sig::dynamic_query::{closure#2}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:620:25
  31:     0x7126a6cf18c5 - <rustc_query_impl[e2adab7f048175ff]::query_impl::fn_sig::dynamic_query::{closure#2} as core[bf256ce745c7f201]::ops::function::FnOnce<(rustc_middle[340db270b268494c]::ty::context::TyCtxt, rustc_span[911f79266521dd07]::def_id::DefId)>>::call_once
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/ops/function.rs:250:5
  32:     0x7126a68d06ac - <rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false> as rustc_query_system[4fa63941cdc3ebb8]::query::config::QueryConfig<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>>::compute
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/lib.rs:113:9
  33:     0x7126a68d06ac - rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr::<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:469:72
  34:     0x7126a68d06ac - rustc_middle[340db270b268494c]::ty::context::tls::enter_context::<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:82:9
  35:     0x7126a68d06ac - <std[5dd5c2799579c4c8]::thread::local::LocalKey<core[bf256ce745c7f201]::cell::Cell<*const ()>>>::try_with::<rustc_middle[340db270b268494c]::ty::context::tls::enter_context<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/thread/local.rs:284:16
  36:     0x7126a68d06ac - <std[5dd5c2799579c4c8]::thread::local::LocalKey<core[bf256ce745c7f201]::cell::Cell<*const ()>>>::with::<rustc_middle[340db270b268494c]::ty::context::tls::enter_context<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/thread/local.rs:260:9
  37:     0x7126a68d06ac - rustc_middle[340db270b268494c]::ty::context::tls::enter_context::<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:79:9
  38:     0x7126a68d06ac - <rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query::<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:151:13
  39:     0x7126a68d06ac - rustc_middle[340db270b268494c]::ty::context::tls::with_related_context::<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:134:9
  40:     0x7126a68d06ac - rustc_middle[340db270b268494c]::ty::context::tls::with_context::<rustc_middle[340db270b268494c]::ty::context::tls::with_related_context<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:112:36
  41:     0x7126a68d06ac - rustc_middle[340db270b268494c]::ty::context::tls::with_context_opt::<rustc_middle[340db270b268494c]::ty::context::tls::with_context<rustc_middle[340db270b268494c]::ty::context::tls::with_related_context<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:101:18
  42:     0x7126a68d06ac - rustc_middle[340db270b268494c]::ty::context::tls::with_context::<rustc_middle[340db270b268494c]::ty::context::tls::with_related_context<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:112:5
  43:     0x7126a68d06ac - rustc_middle[340db270b268494c]::ty::context::tls::with_related_context::<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:125:5
  44:     0x7126a68d06ac - <rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query::<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:136:9
  45:     0x7126a68d06ac - rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr::<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:469:18
  46:     0x7126a68d06ac - rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job::<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt, false>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:405:9
  47:     0x7126a68d06ac - rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::try_execute_query::<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt, false>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:348:13
  48:     0x7126a6c2eb5d - rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::get_query_non_incr::<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:800:32
  49:     0x7126a6c2eb5d - stacker[2c12dcaf2787ac0e]::maybe_grow::<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::get_query_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/.cargo/registry/src/index.crates.io-6f17d22bba15001f/stacker-0.1.15/src/lib.rs:55:9
  50:     0x7126a6c2eb5d - rustc_data_structures[652bf30183ef56b3]::stack::ensure_sufficient_stack::<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::get_query_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_data_structures/src/stack.rs:17:5
  51:     0x7126a6c2eb5d - rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::get_query_non_incr::<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:800:5
  52:     0x7126a6c2eb5d - rustc_query_impl[e2adab7f048175ff]::query_impl::fn_sig::get_query_non_incr::__rust_end_short_backtrace
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:598:26
  53:     0x7126a50274b7 - rustc_middle[340db270b268494c]::query::plumbing::query_ensure::<rustc_query_system[4fa63941cdc3ebb8]::query::caches::DefIdCache<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 24usize]>>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/query/plumbing.rs:172:9
  54:     0x7126a50274b7 - <rustc_middle[340db270b268494c]::query::plumbing::TyCtxtEnsure>::fn_sig::<rustc_span[911f79266521dd07]::def_id::LocalDefId>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/query/plumbing.rs:207:9
  55:     0x7126a50274b7 - rustc_hir_analysis[578b254fcde5e3d0]::collect::lower_item
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/collect.rs:665:13
  56:     0x7126a5017b72 - <rustc_hir_analysis[578b254fcde5e3d0]::collect::CollectItemTypesVisitor as rustc_hir[680d8e47a760be0a]::intravisit::Visitor>::visit_item
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/collect.rs:279:9
  57:     0x7126a5128f2c - rustc_hir_analysis[578b254fcde5e3d0]::check::wfcheck::check_item
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/check/wfcheck.rs:231:5
  58:     0x7126a5128f2c - rustc_hir_analysis[578b254fcde5e3d0]::check::wfcheck::check_well_formed
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_hir_analysis/src/check/wfcheck.rs:195:39
  59:     0x7126a6a23673 - rustc_query_impl[e2adab7f048175ff]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:624:47
  60:     0x7126a6a23673 - rustc_query_impl[e2adab7f048175ff]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2adab7f048175ff]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:548:18
  61:     0x7126a6ca8e4b - rustc_query_impl[e2adab7f048175ff]::query_impl::check_well_formed::dynamic_query::{closure#2}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:620:25
  62:     0x7126a6ca8e4b - <rustc_query_impl[e2adab7f048175ff]::query_impl::check_well_formed::dynamic_query::{closure#2} as core[bf256ce745c7f201]::ops::function::FnOnce<(rustc_middle[340db270b268494c]::ty::context::TyCtxt, rustc_hir[680d8e47a760be0a]::hir_id::OwnerId)>>::call_once
                               at /home/matthias/vcs/github/rust_debug_assertions/library/core/src/ops/function.rs:250:5
  63:     0x7126a698e775 - <rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false> as rustc_query_system[4fa63941cdc3ebb8]::query::config::QueryConfig<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>>::compute
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/lib.rs:113:9
  64:     0x7126a698e775 - rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr::<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_system/src/query/plumbing.rs:469:72
  65:     0x7126a698e775 - rustc_middle[340db270b268494c]::ty::context::tls::enter_context::<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:82:9
  66:     0x7126a698e775 - <std[5dd5c2799579c4c8]::thread::local::LocalKey<core[bf256ce745c7f201]::cell::Cell<*const ()>>>::try_with::<rustc_middle[340db270b268494c]::ty::context::tls::enter_context<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/thread/local.rs:284:16
  67:     0x7126a698e775 - <std[5dd5c2799579c4c8]::thread::local::LocalKey<core[bf256ce745c7f201]::cell::Cell<*const ()>>>::with::<rustc_middle[340db270b268494c]::ty::context::tls::enter_context<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/library/std/src/thread/local.rs:260:9
  68:     0x7126a698e775 - rustc_middle[340db270b268494c]::ty::context::tls::enter_context::<rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:79:9
  69:     0x7126a698e775 - <rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query::<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_query_impl/src/plumbing.rs:151:13
  70:     0x7126a698e775 - rustc_middle[340db270b268494c]::ty::context::tls::with_related_context::<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:134:9
  71:     0x7126a698e775 - rustc_middle[340db270b268494c]::ty::context::tls::with_context::<rustc_middle[340db270b268494c]::ty::context::tls::with_related_context<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:112:36
  72:     0x7126a698e775 - rustc_middle[340db270b268494c]::ty::context::tls::with_context_opt::<rustc_middle[340db270b268494c]::ty::context::tls::with_context<rustc_middle[340db270b268494c]::ty::context::tls::with_related_context<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:101:18
  73:     0x7126a698e775 - rustc_middle[340db270b268494c]::ty::context::tls::with_context::<rustc_middle[340db270b268494c]::ty::context::tls::with_related_context<<rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt as rustc_query_system[4fa63941cdc3ebb8]::query::QueryContext>::start_query<rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[4fa63941cdc3ebb8]::query::plumbing::execute_job_non_incr<rustc_query_impl[e2adab7f048175ff]::DynamicConfig<rustc_query_system[4fa63941cdc3ebb8]::query::caches::VecCache<rustc_hir[680d8e47a760be0a]::hir_id::OwnerId, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2adab7f048175ff]::plumbing::QueryCtxt>::{closure#0}>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>::{closure#0}, rustc_middle[340db270b268494c]::query::erase::Erased<[u8; 1usize]>>
                               at /home/matthias/vcs/github/rust_debug_assertions/compiler/rustc_middle/src/ty/context/tls.rs:112:5
  <snip>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.79.0-dev running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [fn_sig] computing function signature of `mainIterator`
#1 [check_well_formed] checking that `mainIterator` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [fn_sig] computing function signature of `mainIterator`
#1 [check_well_formed] checking that `mainIterator` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate

-->



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"krtab"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

without assertions:
```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for functions
 --> o.rs:3:21
  |
3 | fn mainIterator<_ = _> {
  |                     ^ not allowed in type signatures
  |
help: use type parameters instead
  |
3 | fn mainIterator<TT> {
  |                 ~   ~

error: aborting due to 6 previous errors
```

--- a/compiler/rustc_hir_analysis/src/collect.rs
+++ b/compiler/rustc_hir_analysis/src/collect.rs
@@ -201,12 +201,13 @@ pub(crate) fn placeholder_type_error_diag<'cx, 'tcx>(
         placeholder_types.iter().map(|sp| (*sp, (*type_name).to_string())).collect();
 
     if let Some(generics) = generics {
-        if let Some(arg) = params.iter().find(|arg| {
-            matches!(arg.name, hir::ParamName::Plain(Ident { name: kw::Underscore, .. }))
+        if let Some(span) = params.iter().find_map(|arg| match arg.name {
+            hir::ParamName::Plain(Ident { name: kw::Underscore, span }) => Some(span),
+            _ => None,
         }) {
             // Account for `_` already present in cases like `struct S<_>(_);` and suggest
             // `struct S<T>(T);` instead of `struct S<_, T>(T);`.
-            sugg.push((arg.span, (*type_name).to_string()));
+            sugg.push((span, (*type_name).to_string()));
         } else if let Some(span) = generics.span_for_param_suggestion() {
             // Account for bounds, we want `fn foo<T: E, K>(_: K)` not `fn foo<T, K: E>(_: K)`.
             sugg.push((span, format!(", {type_name}")));

--- a/tests/crashes/123861.rs
+++ b/tests/crashes/123861.rs
@@ -1,5 +1,0 @@
-//@ known-bug: #123861
-//@ needs-rustc-debug-assertions
-
-struct _;
-fn mainIterator<_ = _> {}

--- a/tests/ui/generics/overlapping-errors-span-issue-123861.rs
+++ b/tests/ui/generics/overlapping-errors-span-issue-123861.rs
@@ -1,0 +1,8 @@
+fn mainIterator<_ = _> {}
+//~^ ERROR expected identifier, found reserved identifier `_`
+//~| ERROR   missing parameters for function definition
+//~| ERROR   defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions [invalid_type_param_default]
+//~| WARNING this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+//~| ERROR   the placeholder `_` is not allowed within types on item signatures for functions [E0121]
+
+fn main() {}

--- a/tests/ui/generics/overlapping-errors-span-issue-123861.stderr
+++ b/tests/ui/generics/overlapping-errors-span-issue-123861.stderr
@@ -1,0 +1,52 @@
+error: expected identifier, found reserved identifier `_`
+  --> $DIR/overlapping-errors-span-issue-123861.rs:1:17
+   |
+LL | fn mainIterator<_ = _> {}
+   |                 ^ expected identifier, found reserved identifier
+
+error: missing parameters for function definition
+  --> $DIR/overlapping-errors-span-issue-123861.rs:1:23
+   |
+LL | fn mainIterator<_ = _> {}
+   |                       ^
+   |
+help: add a parameter list
+   |
+LL | fn mainIterator<_ = _>() {}
+   |                       ++
+
+error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions
+  --> $DIR/overlapping-errors-span-issue-123861.rs:1:17
+   |
+LL | fn mainIterator<_ = _> {}
+   |                 ^^^^^
+   |
+   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+   = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
+   = note: `#[deny(invalid_type_param_default)]` on by default
+
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for functions
+  --> $DIR/overlapping-errors-span-issue-123861.rs:1:21
+   |
+LL | fn mainIterator<_ = _> {}
+   |                     ^ not allowed in type signatures
+   |
+help: use type parameters instead
+   |
+LL | fn mainIterator<T = T> {}
+   |                 ~   ~
+
+error: aborting due to 4 previous errors
+
+For more information about this error, try `rustc --explain E0121`.
+Future incompatibility report: Future breakage diagnostic:
+error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions
+  --> $DIR/overlapping-errors-span-issue-123861.rs:1:17
+   |
+LL | fn mainIterator<_ = _> {}
+   |                 ^^^^^
+   |
+   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+   = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
+   = note: `#[deny(invalid_type_param_default)]` on by default
+

